syntax = "proto3";

package comunication;

message ModelRequest {
  string name = 1;        // "resnet18-ibn-v1"
  string version = 2;     // ex.: "1.0.0"
}

message ModelChunk { bytes data = 1; }

message ModelHeader {
  string model_name = 1;   // ex.: "resnet18-ibn-v1"
  string version    = 2;   // ex.: "1.0.7"
  string arch       = 3;   // ex.: "resnet18"
  uint64 total_size = 4;   // bytes do state_dict
  string sha256     = 5;   // integridade dos bytes
}

message ModelStream {
  oneof part {
    ModelHeader header = 1;
    ModelChunk chunk = 2;
  }
}

message ModelUpdateRequest {
  string  client_id    = 1; // "client-3"
  uint64  num_samples  = 2; // peso FedAvg
  bytes   state_bytes  = 3; // torch.save(state_dict)
  string  base_version = 4; // vers√£o do global usada no treino local
}

message ModelUpdateAck {
  bool ok = 1;
  string server_version = 2;
  string msg = 3;
}

service ModelDistributor {
  rpc GetModel (ModelRequest) returns (stream ModelStream);
  rpc SubmitUpdate (ModelUpdateRequest) returns (ModelUpdateAck);
}
